### PR TITLE
revset: resolve "@" symbol's workspace id earlier (while parsing)

### PR DESCRIPTION
**Pull request by martinvonz**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

File: lib/src/revset.rs [hunk 1]
```diff
@@ -880,34 +891,46 @@ fn parse_symbol_rule(
...
+            } else {
+                return Err(RevsetParseError::new(
+                    RevsetParseErrorKind::WorkingCopyWithoutWorkspace,
+                ));
+            }
+        } else {
+            WorkspaceId::new(symbol.strip_suffix('@').unwrap().to_string())
+        };
+        Ok(RevsetExpression::working_copy(workspace_id))
+    } else {
```
Comment (yuja) [inline lib/src/revset.rs:932]: Unfortunately, interpreting symbol here breaks queries like `author("martinvonz@")`.

I think alias expansion and string argument parsing should be performed at higher (closer-to-AST) level, but they aren't right now. Perhaps, the easiest workaround is to set "don't interpret symbol" flag to `ParseState` by `parse_function_argument_to_string_pattern`/`as_literal()`.

File: lib/src/revset.rs [hunk 2]
```diff
@@ -1905,20 +1928,18 @@ impl SymbolResolver for FailingSymbolResolver {
 
 pub type PrefixResolver<'a, T> = Box<dyn Fn(&dyn Repo, &HexPrefix) -> PrefixResolution<T> + 'a>;
 
-/// Resolves the "root" and "@" symbols, branches, remote branches, tags, git
+/// Resolves the "root symbol, branches, remote branches, tags, git
```
Comment (yuja) [inline lib/src/revset.rs:1931]: nit: `"root"`

File: lib/src/revset.rs [hunk 3]
```diff
@@ -2131,7 +2141,7 @@ fn resolve_symbols(
 /// return type `ResolvedExpression` is stricter than `RevsetExpression`,
 /// and isn't designed for such transformation.
 fn resolve_visibility(repo: &dyn Repo, expression: &RevsetExpression) -> ResolvedExpression {
-    // If we add "operation" scope (#1283), visible_heads might be translated to
+    // If we add "#operation" scope (#1283), visible_heads might be translated to
```
Comment (yuja) [inline lib/src/revset.rs:2144]: nit: unrelated change?

File: lib/src/revset.rs (hunk 1 above)
Comment (martinvonz) [reply lib/src/revset.rs:932]: Good catch! I'm not sure how I think a plain `"@"` should be interpreted. We currently interpret it as `@`, but maybe we should interpret it as a (non-special) symbol instead.

I'm starting to think that this PR won't help you at all because you're going to do it in a better way anyway. I really don't mind dropping it, but I'm also happy to add that workaround if it you think it would help you.

Comment (yuja) [reply lib/src/revset.rs:932]: > I'm not sure how I think a plain `"@"` should be interpreted. We currently interpret it as `@`, but maybe we should interpret it as a (non-special) symbol instead.

I agree. Quoted `"@"` can be interpreted as a branch or tag name.

I haven't started any work for #2101 or #2095 yet. Maybe we can go one step towards them by restricting `@` and `workspace@` to be unquoted (i.e. `Rule::identifier`)? `author("@")` requires quoted string instead (if the pattern ends with `@`.)

I think this PR is nice, and I'll probably add `branch@remote` handling on top.

Comment (martinvonz) [reply lib/src/revset.rs:932]: > Maybe we can go one step towards them by restricting `@` and `workspace@` to be unquoted (i.e. `Rule::identifier`)? `author("@")` requires quoted string instead (if the pattern ends with `@`.)

Sounds good. I'll try to update this PR to do that.

Comment (martinvonz) [reply lib/src/revset.rs]: I think that's now done. Let me know if it's not what you had in mind.

**Review by yuja — APPROVED** [review]
LGTM, thanks!
